File: targets/jsts-web/src/harness/tracker.ts
// HARNESS (out of scan scope). Records which planted vulns were exploited at runtime.
import { CATALOG } from "./catalog";

type State = { exploited: boolean; detail: string; at: number };

class Tracker {
  states: Record<string, State> = {};
  constructor() { CATALOG.forEach(([id]) => (this.states[id] = { exploited: false, detail: "", at: 0 })); }
  mark(id: string, detail = ""): void {
    const s = (this.states[id] ||= { exploited: false, detail: "", at: 0 });
    s.exploited = true;
    if (detail) s.detail = detail.slice(0, 300);
    s.at = Date.now();
  }
  reset(): void { Object.values(this.states).forEach((s) => Object.assign(s, { exploited: false, detail: "", at: 0 })); }
  snapshot() {
    const items = CATALOG.map(([id, cwe, dclass, title, category, entry]) => ({
      id, cwe, dclass, title, category, entry, ...(this.states[id] || { exploited: false, detail: "", at: 0 }),
    }));
    return { total: CATALOG.length, exploited: Object.values(this.states).filter((s) => s.exploited).length, items };
  }
}

export const TRACKER = new Tracker();

File: targets/python-web/harness/tracker.py
# HARNESS (out of scan scope). Records which planted vulns were exploited at runtime.
import time
from harness.catalog import CATALOG


class Tracker:
    def __init__(self):
        self.states = {c[0]: {"exploited": False, "detail": "", "at": 0} for c in CATALOG}

    def mark(self, vid, detail=""):
        s = self.states.setdefault(vid, {"exploited": False, "detail": "", "at": 0})
        s["exploited"] = True
        if detail:
            s["detail"] = detail[:300]
        s["at"] = int(time.time() * 1000)

    def reset(self):
        for s in self.states.values():
            s.update(exploited=False, detail="", at=0)

    def snapshot(self):
        items = []
        for cid, cwe, dclass, title, category, entry in CATALOG:
            s = self.states.get(cid, {"exploited": False, "detail": "", "at": 0})
            items.append({"id": cid, "cwe": cwe, "dclass": dclass, "title": title,
                          "category": category, "entry": entry, **s})
        return {"total": len(CATALOG),
                "exploited": sum(1 for s in self.states.values() if s["exploited"]),
                "items": items}


TRACKER = Tracker()
